File: Code/CH-GEE_main.js
//***************************************************************************************************************
//********************************************* Canopy Height Mapper  *******************************************
//***************************************************************************************************************

var CanopyHeightMapper = function(aoi, year, start_date, end_date,startDateGEDI,endDateGEDI,cloudsTh, quantile, model, mask, gedi_type,
  numTreesRF,varSplitRF,minLeafPopuRF,bagFracRF,maxNodesRF,numTreesGBM,shrGBM,samLingRateGBM,maxNodesGBM,lossGBM,maxNodesCART,minLeafPopCART){
    
    //***************************************************************************************************************
    //  Input Data
    //***************************************************************************************************************
  
    var startDateWithYear = year+"-"+start_date; 
    var endDateWithYear   = year+"-"+end_date;
    
    //***************************************************************************************************************
    //  Importing Area of Insterest (AOI) through drawing or uploading
    //***************************************************************************************************************
  
    var aoi2 = ee.Geometry(ee.FeatureCollection(aoi).geometry())//.geometries()
    var polygonArea = aoi2.area({'maxError': 1});
    polygonArea = (polygonArea.divide(10000).round()).getInfo();
    
    //***************************************************************************************************************
    //  Adjusting the Visualization Settings for the AOI
    //***************************************************************************************************************
    
    if(polygonArea < 2000 ){var scale = 10
    Map.centerObject(aoi, 14)}
    else if(polygonArea >= 2000 & polygonArea < 10000){
      scale = 50
      Map.centerObject(aoi, 12)}
    else if(polygonArea >= 10000 & polygonArea < 20000){
      scale = 100
      Map.centerObject(aoi, 10)}
    else if(polygonArea >= 10000 & polygonArea < 330000){
      scale = 100
      Map.centerObject(aoi, 8);
    }else if(polygonArea >= 330000 & polygonArea < 2200000){ 
      scale = 200
      Map.centerObject(aoi, 8);
    }else if(polygonArea >= 2200000 & polygonArea < 10000000){ 
      scale = 250
      Map.centerObject(aoi, 8);
    }else {scale = 250
    Map.centerObject(aoi, 6);}
    
    //***************************************************************************************************************
    //  Selecting one of the three Forest Masks within the AOI
    //***************************************************************************************************************
  
    var library10 = require("users/calvites1990/CH-GEE:ForForestMasking");
    var FNF = library10.ForestMasking(aoi2,mask);
   
    //***************************************************************************************************************
    //  Selecting Dependent variables
    //***************************************************************************************************************
    //***************************************************************************************************************
    //  GEDI Relative Height (RH) metrics
    //***************************************************************************************************************
    var dataset = ee.ImageCollection("LARSE/GEDI/GEDI02_A_002_MONTHLY")
    var library2 = require("users/calvites1990/CH-GEE:L2A_GEDI_source"); 
    var gedi = library2.ToGEDI(dataset,gedi_type,startDateGEDI,endDateGEDI,quantile,FNF) 
    //***************************************************************************************************************
    //  Selecting Independent variables
    //***************************************************************************************************************
    //***************************************************************************************************************
    //  Harmonized Sentinel-2 MSI: MultiSpectral Instrument, Level-2A
    //***************************************************************************************************************
    var library3 = require("users/calvites1990/CH-GEE:Sentinel2_source");
    var s2 = library3.calculateCompositeClip(year, start_date, end_date, cloudsTh, 20,FNF,aoi2);
    
    //***************************************************************************************************************
    // Global Multi-resolution Terrain Elevation Data 2010
    // Slope from Global Multi-resolution Terrain Elevation Data 2010
    //***************************************************************************************************************
    var dem       = ee.Image("USGS/GMTED2010")
    var Mask      = dem.gt(0);
    var demMasked = dem.mask(Mask).rename('dem');
    var slope     = ee.Terrain.slope(demMasked );
    var aspect    = ee.Terrain.aspect(demMasked );
    //***************************************************************************************************************
    //  Sentinel-1
    //***************************************************************************************************************
    var library6 = require("users/calvites1990/CH-GEE:Sentinel1_source");
    var composite = library6.to_sentinel_filtered(year, start_date,end_date,aoi2,FNF)
    //***************************************************************************************************************
    //  Creating a dataset using dependent and independent variables
    //***************************************************************************************************************
    var merged = s2.select(ee.List.sequence(0,11,1)).addBands(gedi)
    .addBands(demMasked).addBands(slope).addBands(aspect).addBands(composite);
    //***************************************************************************************************************
    //  Application of random sampling in large areas
    //***************************************************************************************************************
    if(scale === 100){
     var cellSize = 4000
    }else{  
     if(polygonArea < 5000){
       cellSize = 100                              
     }else if( polygonArea >= 5000 & polygonArea < 1000000){
       cellSize = 4000;// 4000 // Riducendo questo valore aumenti l'accuratezza
     }else if(polygonArea >= 1000000 & polygonArea < 2000000){ 
       cellSize = 6000 //6000
     }else if(polygonArea >= 2000000 & polygonArea < 3000000){
         cellSize = 6000
         }else{cellSize = 50000;
       }}
      //
     var libraryRS = require("users/calvites1990/CH-GEE:RandomSampling");
     var generatedPoints = libraryRS.generateSamplingSites(aoi2, cellSize, 1, FNF);
     var aoi_buffer = generatedPoints.buffer;
     var aoi_prova = aoi_buffer.geometry();
    //***************************************************************************************************************
    //  Sampling configuration for small and large area (4000 km2 is used as the threshold)
    //***************************************************************************************************************
     if(polygonArea <= 4000){
      var reference = merged.sample({
        region: aoi,
        scale: 10, 
        dropNulls: true,
        numPixels: 1e13, 
        tileScale: 4,
        seed: 0,
        geometries: true});
        }else{ 
          reference = merged.sample({
         region: ee.Geometry(aoi_prova),
         scale: scale, 
         dropNulls: true,
         numPixels: 1e13, 
         tileScale: 16,
         seed: 0,
         geometries: true});
        }
    //***************************************************************************************************************
    //  Splitting the dataset into training and validation sets 
    //***************************************************************************************************************
     reference = reference.randomColumn('random');
     var split = 0.7;
     var training = reference.filter(ee.Filter.lt('random', split));
     var validation = reference.filter(ee.Filter.gte('random', split));
  
    //***************************************************************************************************************
    // Colnames all of used variables
    //***************************************************************************************************************
     var predictorsNames = s2.select(ee.List.sequence(0,11,1))
    .addBands(demMasked).addBands(slope).addBands(aspect)//.bandNames();
    .addBands(composite).bandNames();
    //***************************************************************************************************************
    // Configurating the hyperparameters in each of the three machine learning algorithms:
    // RF   - Random Forest 
    // CART - Classification And Regression Trees classifier
    // GB   - Gradient Tree Boost 
    //***************************************************************************************************************
    if(model=="RF"){
      var  classifier = ee.Classifier.smileRandomForest({
        numberOfTrees:ee.Number(numTreesRF),
        variablesPerSplit: ee.Number(varSplitRF),
        minLeafPopulation: ee.Number(minLeafPopuRF), 
        bagFraction: ee.Number(bagFracRF), 
        maxNodes: ee.Number(maxNodesRF)
        }).setOutputMode("Regression") // used to predict class / continuous variables etc.
                      .train(training, "rh", predictorsNames); 
                      }
    //***************************************************************************************************************
     if(model=="CART"){
     classifier = ee.Classifier.smileCart({
       maxNodes: maxNodesCART,
       minLeafPopulation: ee.Number(minLeafPopCART)
       }).train(training, "rh", predictorsNames)
                      .setOutputMode("Regression"); // used to predict class / continuous variables etc.
        }
    //***************************************************************************************************************
    if(model=="GBM"){
     classifier = ee.Classifier.smileGradientTreeBoost({
       numberOfTrees: ee.Number(numTreesGBM),
      shrinkage: ee.Number(shrGBM),
      samplingRate: ee.Number(samLingRateGBM),
      maxNodes: ee.Number(maxNodesGBM),
     // loss: ee.String(lossGBM)
      }).train(
      training, "rh", predictorsNames)
      .setOutputMode("Regression"); // used to predict class / continuous variables etc.
     }
    //***************************************************************************************************************
    // Prediction of canopy heights 
    //***************************************************************************************************************
    var classified = merged.classify(classifier);
  
    //***************************************************************************************************************
    // Scatter Plot 
    //***************************************************************************************************************
    var library4  = require("users/calvites1990/CH-GEE:ForPlots");
    var validated = validation.classify(classifier);
    var RMSE      = library4.CalculationRMSE(validated);
    library4.SCPLOT(validated,RMSE);
    //***************************************************************************************************************
    // Variable Importance Plot
    //***************************************************************************************************************
    library4.VARIMP(classifier);//* variable important function
    //***************************************************************************************************************
    // Canopy height map
    //***************************************************************************************************************
    library4.scalecolor(0,50,classified);
    //
     return([classified])
    };
  exports.CanopyHeightMapper = CanopyHeightMapper;
  
  //***************************************************** End *****************************************************

File: Pycode/ch_gee_main.py
import ee
import geemap
import numpy as np
from typing import Union, List, Dict, Any

def canopy_height_mapper(
    aoi: ee.Geometry,
    year: int,
    start_date: str,
    end_date: str,
    start_date_gedi: str,
    end_date_gedi: str,
    clouds_th: float,
    quantile: str,
    model: str,
    mask: str,
    gedi_type: str,
    num_trees_rf: int,
    var_split_rf: int,
    min_leaf_pop_rf: int,
    bag_frac_rf: float,
    max_nodes_rf: int,
    num_trees_gbm: int,
    shr_gbm: float,
    sampling_rate_gbm: float,
    max_nodes_gbm: int,
    loss_gbm: str,
    max_nodes_cart: int,
    min_leaf_pop_cart: int
) -> ee.Image:
    """
    Main function for canopy height mapping using Earth Engine.
    
    Args:
        aoi: Area of interest as Earth Engine Geometry
        year: Year for analysis
        start_date: Start date for satellite data
        end_date: End date for satellite data
        start_date_gedi: Start date for GEDI data
        end_date_gedi: End date for GEDI data
        clouds_th: Cloud threshold
        quantile: Quantile for GEDI data
        model: Model type ('RF', 'GBM', or 'CART')
        mask: Mask type ('DW', 'FNF', or 'none')
        gedi_type: GEDI data type ('singleGEDI' or 'meanGEDI')
        num_trees_rf: Number of trees for Random Forest
        var_split_rf: Variable split for Random Forest
        min_leaf_pop_rf: Minimum leaf population for Random Forest
        bag_frac_rf: Bagging fraction for Random Forest
        max_nodes_rf: Maximum nodes for Random Forest
        num_trees_gbm: Number of trees for GBM
        shr_gbm: Shrinkage for GBM
        sampling_rate_gbm: Sampling rate for GBM
        max_nodes_gbm: Maximum nodes for GBM
        loss_gbm: Loss function for GBM
        max_nodes_cart: Maximum nodes for CART
        min_leaf_pop_cart: Minimum leaf population for CART
    
    Returns:
        ee.Image: Classified canopy height map
    """
    # Calculate area
    polygon_area = aoi.area({'maxError': 1})
    polygon_area = polygon_area.divide(10000).round()
    
    # Import required modules
    from .l2a_gedi_source import get_gedi_data
    from .sentinel1_source import get_sentinel1_data
    from .sentinel2_source import get_sentinel2_data
    from .for_forest_masking import apply_forest_mask
    from .random_sampling import create_training_data
    
    # Get input data
    gedi_data = get_gedi_data(aoi, start_date_gedi, end_date_gedi, quantile)
    s1_data = get_sentinel1_data(aoi, year, start_date, end_date)
    s2_data = get_sentinel2_data(aoi, year, start_date, end_date, clouds_th)
    
    # Apply forest mask if needed
    if mask != 'none':
        gedi_data = apply_forest_mask(gedi_data, mask)
    
    # Create training data
    training_data = create_training_data(gedi_data, s1_data, s2_data)
    
    # Train model based on selected type
    if model == 'RF':
        classifier = ee.Classifier.smileRandomForest(
            numberOfTrees=num_trees_rf,
            variablesPerSplit=var_split_rf,
            minLeafPopulation=min_leaf_pop_rf,
            bagFraction=bag_frac_rf,
            maxNodes=max_nodes_rf
        )
    elif model == 'GBM':
        classifier = ee.Classifier.smileGradientTreeBoost(
            numberOfTrees=num_trees_gbm,
            shrinkage=shr_gbm,
            samplingRate=sampling_rate_gbm,
            maxNodes=max_nodes_gbm,
            loss=loss_gbm
        )
    else:  # CART
        classifier = ee.Classifier.smileCart(
            maxNodes=max_nodes_cart,
            minLeafPopulation=min_leaf_pop_cart
        )
    
    # Train the classifier
    trained_classifier = classifier.train(
        features=training_data,
        classProperty='height',
        inputProperties=s1_data.bandNames().cat(s2_data.bandNames())
    )
    
    # Classify the image
    classified = s1_data.addBands(s2_data).classify(trained_classifier)
    
    # Scale the output
    classified = classified.multiply(50).round()
    
    return classified

def get_variable_importance(classifier: ee.Classifier) -> Dict[str, float]:
    """
    Get variable importance from the trained classifier.
    
    Args:
        classifier: Trained Earth Engine classifier
    
    Returns:
        Dict[str, float]: Dictionary of variable names and their importance scores
    """
    return classifier.explain().get('importance').getInfo() 